### DISCUSSION TITLE
[Question] Redirecting logs to layout element

### DISCUSSION THREAD

**Discussion by mayeroa** [Textualize/rich#1143] [Q&A]
First of all, many thanks for this wonderful library. It helps a lot!

I was wondering something. I would like to use rich to follow the training of my deep learning code.
To do so, I really think the following example (https://github.com/willmcgugan/rich/blob/master/examples/fullscreen.py) is of interest of what I want to get.

My question is about the layout and the logging system. Having the progress bars at the bottom to measure the progress of the training is pretty interesting. I can also imagine to report metrics result in a table for visual progression. But I wonder if we can combine a fullscreen application (like in the above example) with some logging. I need to get logs to allow debugging during the training and it would be awesome if I can branch it to an element of the proposed layout.

For instance, would it be possible to redirect console/progress bars/other logs to the **body** element of  the layout? If that is possible, how to proceed?

Thanks in advance for your time!

**Comment by willmcgugan** [comment]
Glad you like Rich!

I have been thinking about a log renderable which would give you a scrolling window of text in a layout. Although it may be of limited use until I've added more dynamic features to make it interactive (keyboard/mouse scroll).

> For instance, would it be possible to redirect console/progress bars/other logs to the body element of the layout? If that is possible, how to proceed?

There's no explicit support for that at the moment, although I'm sure something along those lines is possible.

In the meantime, if you want a fullscreen display you might want to do your logging with the Python `logging` module and configure it to log to a file. In another window you can tail that log file.

**Reply by mayeroa** [reply]
Thanks for your detailed answer.
Concerning the scroll, that was my next question but you did anticipate. Having such dynamic functionalities would be awesome to have and build richer console interfaces. Do you have any idea when such work will start?

> In the meantime, if you want a fullscreen display you might want to do your logging with the Python `logging` module and configure it to log to a file. In another window you can tail that log file.

I will have a look into that, it should perfectly do the trick for now.

Thx!

**Reply by willmcgugan** [reply]
> Do you have any idea when such work will start?

It's kind of underway at the moment. But its a large undertaking, and this is a part time project, so I can't promise any timescale...

**Reply by mayeroa** [reply]
No problem. Many thanks by he way. :)

**Reply by sergiors** [reply]
Is there a way to remove top row in table? I guess it's a good way to do a tail log for now

**Reply by jakane** [reply]
I would like to second this request. I do need something that I can treat as a console for random messages.

So far I have implemented it as a Layout of a specific size, then I create a fixed-length array with:
``` python
from collections import deque
from rich.layout import Layout
from rich.live import Live
from time import sleep

console_messages = deque(maxlen=10)
layout = Layout()
layout.split_column(Layout(name="console"), Layout(name="stats"))
layout["console"].size = 10

with Live(layout) as live:
  #... some processing
  for i in range(1,21):
    console_messages.append(f"Here is a message {i}")
    layout["console"].update("\n".join(console_messages))
    sleep(0.2)
```
It works, but having something like:
``` python
layout["console"].print("Here is a message")
```
would be much lighter weight and more intuitive for me.

**Reply by sjoerdtimmer** [reply]
> It's kind of underway at the moment. But its a large undertaking, and this is a part time project, so I can't promise any timescale...

Cool. Is there by now anything we can experiment with?

I'm currently just writing messages to a textual TextLog but I would love to have a rich.Console that is renderable or a logging.Handler derivative that is renderable. Either way: something that catches logging messages that can be embedded in a textual app would be a cool thing to have.
